### PR TITLE
Add MarshalBinary/UnmarshalBinary support

### LINKED ISSUE #67: sha256-simd does not implement encoding.BinaryMarshaler/encoding.BinaryUnmarshaler
Hi!

I know this was discussed in #29 and decided against, but the lack of implementation of these interfaces makes this library not a complete drop-in replacement for crypto/sha256.

Would it be possible to implement these?

### PR DESCRIPTION
**Pull request by klauspost**
Fixes #67 

Mainly copied from stdlib.

### CODE REVIEW

**Comment by SuperSandro2000** [comment]
Can you create a new release with this fix? On 11th gen Intel CPUs which have avx512 it causes a compilation error.

**Comment by klauspost** [comment]
@SuperSandro2000 v1.0.1 released.

File: sha256.go [hunk 1]
```diff
@@ -386,3 +387,82 @@ var _K = []uint32{
...
+	return b, nil
+}
+
+func (d *digest) UnmarshalBinary(b []byte) error {
+	if len(b) < len(magic256) || string(b[:len(magic256)]) != magic256 {
+		return errors.New("crypto/sha256: invalid hash state identifier")
+	}
+	if len(b) != marshaledSize {
+		return errors.New("crypto/sha256: invalid hash state size")
+	}
```
Comment (stefansundin) [inline sha256.go:419]: Perhaps these errors shouldn't be referencing `crypto/sha256`?

Comment (klauspost) [reply sha256.go:419]: Feel free to send a PR.

**Pull request merged**